### PR TITLE
Themed error page redux

### PR DESCRIPTION
**Pull request by MrPetovan**
Part 2 of the breakdown of #7050 into smaller PRs.

Builds on #7067, I'll release this PR when the other will be merged to reduce the review noise.

### CODE REVIEW

**Comment by nupplaphil** [comment]
#7067 merged :-) Ready for this changes :D

**Comment by MrPetovan** [comment]
Oh look, conflicts!

**Comment by MrPetovan** [comment]
Not too bad considering we were working on the same files in parallel.

File: mod/display.php [hunk 1]
```diff
@@ -76,7 +77,7 @@ function display_init(App $a)
 	}
 
 	if (!DBA::isResult($item)) {
-		System::httpExit(404);
+		return;
```
Comment (nupplaphil) [inline mod/display.php]: Why did you remove this exception?

File: src/Module/Photo.php [hunk 2]
```diff
@@ -74,9 +75,7 @@ public static function init()
 		}
 
 		if ($photo === false) {
-			// not using System::httpExit() because we don't want html here.
-			header($_SERVER["SERVER_PROTOCOL"] . " 404 Not Found" , true, 404);
-			exit();
+			System::httpExit('404', 'Not Found');
```
Comment (nupplaphil) [inline src/Module/Photo.php]: Why don't we use an exception too?

File: mod/videos.php [hunk 3]
```diff
@@ -84,33 +84,6 @@ function videos_post(App $a)
 	}
 
 	if (($a->argc == 2) && !empty($_POST['delete']) && !empty($_POST['id'])) {
-		// Check if we should do HTML-based delete confirmation
-		if (empty($_REQUEST['confirm'])) {
```
[the remark this answers is not in the document]
Comment (MrPetovan) [reply mod/videos.php]: The `/videos` module interactions are just not implemented. There's no link/form that can bring you to this part of the code.

File: src/App.php [hunk 4]
```diff
@@ -1127,123 +1138,115 @@ public function runFrontend()
 			'title' => ''
 		];
 
-		if (strlen($this->module)) {
-			// Compatibility with the Android Diaspora client
-			if ($this->module == 'stream') {
-				$this->internalRedirect('network?f=&order=post');
-			}
+		// Compatibility with the Android Diaspora client
```
[the remark this answers is not in the document]
Comment (MrPetovan) [reply src/App.php]: Yes, with a redirection in the modules. The Router only allows for URL rewriting so far.

File: src/App.php [hunk 5]
```diff
@@ -1310,13 +1314,6 @@ public function runFrontend()
 		 */
 		$this->initFooter();
 
-		/* now that we've been through the module content, see if the page reported
-		 * a permission problem and if so, a 403 response would seem to be in order.
-		 */
-		if (stristr(implode("", $_SESSION['sysmsg']), Core\L10n::t('Permission denied'))) {
```
[the remark this answers is not in the document]
Comment (MrPetovan) [reply src/App.php]: It was ineffective (comparing user messages with a translated string) and ultimately useless because we have proper Forbidden error pages now.

File: mod/display.php (hunk 1 above)
Comment (MrPetovan) [reply mod/display.php]: Any exception thrown in the `init()` method would produce a non-styled error page, which avoiding them was the initial goal of this PR.

The item availability is re-checked in the `display_content()` function anyway, and then a third time in the same function when we finally factor in the logged in user permissions.
I started refactoring the display module in `src/Module` but I had to stop because of how byzantine the logic is.

File: src/Module/Photo.php (hunk 2 above)
Comment (MrPetovan) [reply src/Module/Photo.php]: This module output image data, so it doesn't make sense to output HTML. It's the only other place in the code where calling `System::httpExit()` is actually warranted.

File: src/App.php [hunk 6]
```diff
@@ -1011,19 +1010,28 @@ public function runFrontend()
 	{
 		// Missing DB connection: ERROR
 		if ($this->getMode()->has(App\Mode::LOCALCONFIGPRESENT) && !$this->getMode()->has(App\Mode::DBAVAILABLE)) {
-			Core\System::httpExit(500, ['title' => 'Error 500 - Internal Server Error', 'description' => 'Apologies but the website is unavailable at the moment.']);
+			echo Module\Special\HTTPException::rawContent(
```
[the remark this answers is not in the document]
Comment (MrPetovan) [reply src/App.php]: The problem I currently have for this approach is that this special module uses templating. At the moment we don't seem to use config values in the current template engine but it may happen. I'm afraid of circular dependency between the error module and the templating engine in case the App isn't properly initialized.

If we ignore that, the other question I had is if I could have a try/catch block inside another. Here's my ideal structure that I don't know is feasible:
```php
class App {
    ...
    public function runFrontend() {
        try {
            // Missing DB connection: ERROR
            if ($this->getMode()->has(App\Mode::LOCALCONFIGPRESENT) && !$this->getMode()->has(App\Mode::DBAVAILABLE)) {
                throw new HTTPException\InternalServerErrorException('Apologies but the website is unavailable at the moment.');
            }
           
            ...
           
            try {
                // Processing module content
            } catch(HTTPException $e) {
                $content = Module\Special\HTTPException::content($e);
            }
           
            ...
           
            require_once $template;
        } catch(HTTPException $e) {
            $content = Module\Special\HTTPException::rawContent($e);
            System::httpExit($e->getCode(), $e->httpdesc, $content);
        }
    }
    ...
}
```

[the remark this answers is not in the document]
Comment (nupplaphil) [reply src/App.php]: hm .. I see a more "radical" approach for the future:

index.php:
```php
$dependencyManager = new DependencyManager(Dice $diceDILibrary);
$router = new Router($dependencyManager);

$module = $router->getModule();

$module->run();
```

Router:
```php
class Router {
     /**
      * @return IModule
      */
    public function getModule()
    { 
        try {
           $moduleName = $this->getModuleBecauseOfRoutes();
       
           // the dependency manager creates the IModule for us with all dependencies (including App, DBA, ...)
           return $this->dependencyManager->create($moduleName);
         } catch (Exception $e) {
             // This is a system, non-themed exception
         }
    }
}
```

and a module looks like:

```php
class Feed implements IModule
{
          public function __construct(App $app, OStatus $oStatus) 
          {
               // the constructor is called per $this->dependencyManager->create() and all dependencies are automatically "here"
               [...]
          }

         public function run()
         {
                try
                {
                   $app = $this->app;
                   $oStatus = $this->oStatus;
                } catch (HTTPException $e) {
                     // this is a themed, well defined exception
               }  
         } 
}
```

[the remark this answers is not in the document]
Comment (nupplaphil) [reply src/App.php]: It's called https://en.wikipedia.org/wiki/Inversion_of_control

[the remark this answers is not in the document]
Comment (MrPetovan) [reply src/App.php]: I like the idea of having an `App`-like class with a method that's the single call in the app entry points. Thats's what `Friendica\Core\Console` is about  (and what `bin/worker.php` and `bin/daemon.php` aren't) and I love this kind of simplicity.

In your example, it feels to me that the Router would de facto take that role with `getModule()` as the new `runFrontend()`, and I'd be fine with that.

[the remark this answers is not in the document]
Comment (nupplaphil) [reply src/App.php]: hm .. Console would be more like

console.php
```php
$dependencyManager = new DependencyManager(Dice $dependencyDILibrary);

$console = $dependencyManager->create($cmdName);

// in fact, this is exactly what console is currently doing like I thought about module
$console->execute();
```


```php
class AutomaticInstallationConsole extends Console
{
        public function __construct(App $app)
        {
        }
}
```

Because we now have an issue, where we must not load App, but with the current structure, we cannot skip this step and load everything.

The difference is that `getModule()` doesn't need any dependency except the route to the module. The current `App->runFrontend()` needs a fully compiled dependency structure (see the current constructor).

[the remark this answers is not in the document]
Comment (MrPetovan) [reply src/App.php]: Yep, sounds about right.

**Comment by nupplaphil** [comment]
solves #7047
